### DISCUSSION TITLE
Modify and save database after simulation

### DISCUSSION THREAD

**Discussion by krabo0om** [ucb-bar/hammer#783] [Support (Q&A)] [answered]
I want to convert the `.vpd` files from a `sim-debug` (from VCS) to gziped `.vcd` files (for Voltus). Xcelium does not appear to be supported yet for gate-level simulation. 

I have inserted an additional step after VCS's `run_simulation` to manipulate the waveforms that are given in `sim.outputs.waveforms`. As a result, their path/name changes. To reflect the changed path, I would like to update the `sim.outputs.waveforms` variable. First, for consistency and, second, to provide the correct paths for power analysis later. 

However, when I use `x.set_setting("sim.outputs.waveforms", new_waveform_paths)` in my hook, it does not change the `sim-debug-inputs.json` file. 

How can I make my changes permanent or is there a better way?

**Comment by harrisonliew** [comment]
If I'm not mistaken, I think you're using Hammer within Chipyard. A couple answers:

- Xcelium does support gate-level simulation, but it just hasn't been enabled in Chipyard yet.
- You could (and should) write to FSDB instead of VPD and converting to VCD. Just make sure the VCS version you are using isn't too new (see #769).
- The VCS plugin does not write the `sim.outputs.waveforms` key, as you may have found. Instead, it is set within Chipyard in [sim.mk here](https://github.com/ucb-bar/chipyard/blob/352cc773b55c509245aa4755450fa88130dba6a2/vlsi/sim.mk#L68-L70). You will need to modify your sim.mk instead if you want to change the waveform file.

**Reply by krabo0om** [reply]
Thanks, yes, I'm using Hammer within Chipyard.
 
>     * Xcelium does support gate-level simulation, but it just hasn't been enabled in Chipyard yet.

Would that be easy to do for someone rather unfamiliar with Chipyard/Hammer like me? Would be the best solution. 
 
>     * You could (and should) write to FSDB instead of VPD and converting to VCD. Just make sure the VCS version you are using isn't too new (see Investigate VPD, FSDB Support #769 ).

This exact issue was to motivation to convert it manually. VCS from 2019 appears to be too new ("empty" fdsb file according to voltus) and our version from 2017 is not a full 64bit installation. I can ask for 2018 to be installed. I am also limited to Voltus 2021 because 2022 crashes when reading `ChipTop_power_constraints.tcl` (although created by Innovus of the same version). 

>     * The VCS plugin does not write the `sim.outputs.waveforms` key, as you may have found. Instead, it is set within Chipyard in [sim.mk here](https://github.com/ucb-bar/chipyard/blob/352cc773b55c509245aa4755450fa88130dba6a2/vlsi/sim.mk#L68-L70). You will need to modify your sim.mk instead if you want to change the waveform file.

That works. However, VCS still writes in the vpd format. Because my hook could be executed more than once, I have to check if the `.vcd.gz` file is actually a vpd or has already been converted to a vcd. Not a big problem, but also not "clean". 
Currently, I'm using a hook before the power command to update the database and replace the `power.input.waveforms` paths with `.vcd.gz`.

**Reply by harrisonliew** [reply]
> Would that be easy to do for someone rather unfamiliar with Chipyard/Hammer like me? Would be the best solution.

This will require that you dive into some Makefiles and understand how the [Xcelium plugin](https://github.com/ucb-bar/hammer/tree/master/hammer/sim/xcelium) works. If you want to try, you need to create a Xcelium-specific version of [sim.mk](https://github.com/ucb-bar/chipyard/blob/main/vlsi/sim.mk) using [this Makefile](https://github.com/ucb-bar/chipyard/blob/main/sims/xcelium/Makefile) as the reference for how to set the `sim.` keys. We just haven't gotten around to doing it yet, so if you are successful, a PR to Chipyard would be appreciated.

> However, VCS still writes in the vpd format.

You probably have the `USE_VPD` Make variable set. Without it, it would generate FSDB by default. Anyways this doesn't matter for you since you can't use FSDB.

> Because my hook could be executed more than once, I have to check if the .vcd.gz file is actually a vpd or has already been converted to a vcd.

Is there a reason why the hook could be executed more than once? Is this due to your own flow, or some Hammer issue?

> Currently, I'm using a hook before the power command to update the database and replace the power.input.waveforms paths with .vcd.gz.

This is a good workaround for now.

**Comment by krabo0om** [comment, marked as answer]
> You probably have the USE_VPD Make variable set. Without it, it would generate FSDB by default.

I tried all the combinations :) 

> Is there a reason why the hook could be executed more than once? Is this due to your own flow, or some Hammer issue?

I think it is due to my flow. Maybe it was also just an artifact of me testing the method repeatedly. In any case, the current workaround is good enough for now.


For completeness, I put the code here if anybody experiences the same problem:

```python
import pathlib

def vpd2vcd_gz(x: HammerTool) -> bool:
    try:
      waveforms = x.get_setting("sim.outputs.waveforms")
    except KeyError:
        return True
    for waveform in waveforms:
        waveform = pathlib.Path(waveform)
        if waveform.suffix == '.vpd' and waveform.exists():
            vcd = waveform.with_suffix('.vcd')
            subprocess.run(['vcsplit', '-o', vcd, '-scope', x.get_setting('sim.inputs.tb_dut'), '-full64', waveform])
            subprocess.run(['gzip', vcd])
            waveform.unlink()
    return True

def vpd2vcd_path_fixes(x: HammerTool) -> bool:
    x.waveforms = [waveform.removesuffix('vpd') + ('vcd.gz') for waveform in x.waveforms]
    return True

class ExampleDriver(CLIDriver):
    def get_extra_sim_hooks(self) -> List[HammerToolHookAction]:
        extra_hooks = [
            HammerTool.make_post_insertion_hook("run_simulation", vpd2vcd_gz),
        ]   
        return extra_hooks

    def get_extra_power_hooks(self) -> List[HammerToolHookAction]:
        extra_hooks = [
            HammerTool.make_pre_insertion_hook("init_technology", vpd2vcd_path_fixes),
        ]   
        return extra_hooks
```

If you are using Chipyard, merge the code with the existing example-vlsi.
